streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -342,7 +379,18 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         maybeThrowTaskExceptions(taskCloseExceptions);
 
-        createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        final Collection<Task> newActiveTasks = createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        final Set<Task> activeTasksNeedCommit;
+        // Find new active tasks which need commit.
+        if (newActiveTasks == null || newActiveTasks.isEmpty()) {
+            activeTasksNeedCommit = new HashSet<>();
+        } else {
+            activeTasksNeedCommit = newActiveTasks.stream().filter(Task::commitNeeded).collect(Collectors.toSet());
```

Comment:
For newly created tasks, they should not have `commitNeeded` flag set. I think @ableegoldman 's comment was that the conditions are `(a) EOS is used (since this is a txn-based issue to begin with) and (b) if we have newly-added active tasks (as discussed previously)`, but maybe `(c) there is already a txn ongoing)`.

Note that for EOS-v1, and EOS-v2, condition (c) is different (and that's why I was originally thinking to wait until we only have EOS-v2, but after some thinking I'm feeling more towards narrowing the scope of fix to EOS-v2 since it is the commonly used config now, and fixing EOS-v1 may incur unnecessary complexity with little value). For EOS-v2, we just need to check `threadProducer.hasInflighTxn`.

Reply:
I see. I have modified this check accordingly, ie now I am checking if there are newly created active tasks and there's an in-flight txn, then look to commit tasks. Since `threadProducer()` works only for EOS-v2, this fix implicitly applies only for that version.